Tests/PostgresNIOTests/New/PSQLRowStreamTests.swift

```diff
@@ -231,6 +240,196 @@ final class PSQLRowStreamTests: XCTestCase {
+        XCTAssertEqual(dataSource.hitDemand, 1)
+
+        stream.receive([
+            [ByteBuffer(string: "2")],
+            [ByteBuffer(string: "3")]
+        ])
+        XCTAssertEqual(dataSource.hitDemand, 2)
+
+        stream.receive([
+            [ByteBuffer(string: "4")],
+            [ByteBuffer(string: "5")]
+        ])
+        XCTAssertEqual(dataSource.hitDemand, 3)
+
+        stream.receive(completion: .success("SELECT 2"))
+
+        XCTAssertNoThrow(try future.wait())
+    }
+
+    func testDrainBeforeStreamHasFinishedWhenThereIsAlreadyAConsumer() {
+        let dataSource = CountingDataSource()
+        let stream = PSQLRowStream(
+            source: .stream(
+                [self.makeColumnDescription(name: "foo", dataType: .text, format: .binary)],
+                dataSource
+            ),
+            eventLoop: self.eventLoop,
+            logger: self.logger
+        )
+        XCTAssertEqual(dataSource.hitDemand, 0)
+        XCTAssertEqual(dataSource.hitCancel, 0)
+
+        stream.receive([
+            [ByteBuffer(string: "0")],
+            [ByteBuffer(string: "1")]
+        ])
+
+        XCTAssertEqual(dataSource.hitDemand, 0, "Before we have a consumer demand is not signaled")
+
+        // attach consumers
+        let allFuture = stream.all()
+        XCTAssertEqual(dataSource.hitDemand, 1)
+        let drainFuture = stream.drain()
+        XCTAssertEqual(dataSource.hitDemand, 2)
+
+        stream.receive([
+            [ByteBuffer(string: "2")],
+            [ByteBuffer(string: "3")]
+        ])
+        XCTAssertEqual(dataSource.hitDemand, 3)
+
+        stream.receive([
+            [ByteBuffer(string: "4")],
+            [ByteBuffer(string: "5")]
+        ])
+        XCTAssertEqual(dataSource.hitDemand, 4)
+
+        stream.receive(completion: .success("SELECT 2"))
+
+        XCTAssertNoThrow(try drainFuture.wait())
+
+        var rows: [PostgresRow]?
+        XCTAssertNoThrow(rows = try allFuture.wait())
+        XCTAssertEqual(rows?.count, 6)
+    }
+
+    func testDrainBeforeStreamHasFinishedWhenThereIsAlreadyAnAsyncConsumer() {
+        let dataSource = CountingDataSource()
+        let stream = PSQLRowStream(
+            source: .stream(
+                [self.makeColumnDescription(name: "foo", dataType: .text, format: .binary)],
+                dataSource
+            ),
+            eventLoop: self.eventLoop,
+            logger: self.logger
+        )
+        XCTAssertEqual(dataSource.hitDemand, 0)
+        XCTAssertEqual(dataSource.hitCancel, 0)
+
+        stream.receive([
+            [ByteBuffer(string: "0")],
+            [ByteBuffer(string: "1")]
+        ])
+
+        XCTAssertEqual(dataSource.hitDemand, 0, "Before we have a consumer demand is not signaled")
+
+        // attach consumers
+        let rowSequence = stream.asyncSequence()
+        XCTAssertEqual(dataSource.hitDemand, 0)
+        let drainFuture = stream.drain()
+        XCTAssertEqual(dataSource.hitDemand, 1)
```

Comment:
This might happen when someone is using the new `query` function. an async-seq is attached but user might simply ignore it.